File: archive/frontend-server-side-recommendation/recommendation/intent.ts
import { z } from "zod";
import { ChatOpenAI } from "@langchain/openai";
import { env } from "@/lib/env";
import { filterKnownTags } from "@/lib/recommendation/all-tags";
import type { ParsedUserIntent } from "@/lib/types";

const parsedIntentSchema = z.object({
  preferred_tags: z.array(z.string()).default([]),
  avoid_tags: z.array(z.string()).default([]),
  free_text_intent: z.string().default(""),
  constraints: z
    .object({
      price_max: z.number().optional(),
      year_min: z.number().optional(),
      single_player: z.boolean().optional(),
      multiplayer: z.boolean().optional(),
    })
    .partial()
    .optional(),
});

function extractNumbers(prompt: string) {
  const priceMatch = prompt.match(/\$?(\d{1,3})(?:\s*(?:dollars|usd|bucks|under|max))/i);
  const yearMatch = prompt.match(/(?:after|since|newer than)\s+(20\d{2}|19\d{2})/i);

  return {
    price_max: priceMatch ? Number(priceMatch[1]) : undefined,
    year_min: yearMatch ? Number(yearMatch[1]) : undefined,
  };
}

function heuristicIntent(prompt: string): ParsedUserIntent {
  const lowerPrompt = prompt.toLowerCase();
  const matches = new Set<string>();

  for (const tag of filterKnownTags([lowerPrompt])) {
    matches.add(tag);
  }

  // Phrase-level scan for known Steam tags.
  const phrases = lowerPrompt.split(/[,.!?]/).flatMap((segment) => [
    segment.trim(),
    ...segment.trim().split(/\s+/).reduce<string[]>((acc, word, index, parts) => {
      const two = `${word} ${parts[index + 1] ?? ""}`.trim();
      const three = `${word} ${parts[index + 1] ?? ""} ${parts[index + 2] ?? ""}`.trim();
      if (two) acc.push(two);
      if (three) acc.push(three);
      return acc;
    }, []),
  ]);

  for (const candidate of phrases) {
    const filtered = filterKnownTags([candidate]);
    if (filtered[0]) {
      matches.add(filtered[0]);
    }
  }

  const avoidParts = lowerPrompt.split(/but not|avoid|no |don't want|not /i);
  const avoidTags =
    avoidParts.length > 1 ? filterKnownTags(avoidParts.slice(1).join(" ").split(/\s+/)) : [];

  return {
    preferred_tags: [...matches].filter((tag) => !avoidTags.includes(tag)),
    avoid_tags: avoidTags,
    free_text_intent: prompt.trim(),
    constraints: {
      ...extractNumbers(prompt),
      single_player: /single[\s-]?player/i.test(prompt) ? true : undefined,
      multiplayer: /multiplayer|co-op|coop|friends/i.test(prompt) ? true : undefined,
    },
  };
}

export async function parseUserIntent(prompt: string) {
  if (!env.openAiApiKey) {
    return heuristicIntent(prompt);
  }

  const model = new ChatOpenAI({
    apiKey: env.openAiApiKey,
    model: "gpt-4.1-mini",
    temperature: 0,
  });

  const structuredModel = model.withStructuredOutput(parsedIntentSchema);
  const result = await structuredModel.invoke([
    {
      role: "system",
      content:
        "You extract Steam search intent. Only include tags that are exact Steam tags. Put non-tag nuance in free_text_intent.",
    },
    {
      role: "user",
      content: prompt,
    },
  ]);

  return {
    preferred_tags: filterKnownTags(result.preferred_tags ?? []),
    avoid_tags: filterKnownTags(result.avoid_tags ?? []),
    free_text_intent: result.free_text_intent || prompt.trim(),
    constraints: result.constraints,
  } satisfies ParsedUserIntent;
}

File: backend/app/services/recommendation/intent.py
from __future__ import annotations

import re

from langchain_openai import ChatOpenAI

from app.core.config import get_settings
from app.schemas.recommendation import IntentConstraints, ParsedUserIntent
from app.services.recommendation.all_tags import filter_known_tags


def extract_numbers(prompt: str) -> dict[str, int | None]:
    price_match = re.search(r"\$?(\d{1,3})(?:\s*(?:dollars|usd|bucks|under|max))", prompt, re.I)
    year_match = re.search(r"(?:after|since|newer than)\s+(20\d{2}|19\d{2})", prompt, re.I)
    return {
        "price_max": int(price_match.group(1)) if price_match else None,
        "year_min": int(year_match.group(1)) if year_match else None,
    }


def heuristic_intent(prompt: str) -> ParsedUserIntent:
    lower_prompt = prompt.lower()
    matches: set[str] = set()

    matches.update(filter_known_tags([lower_prompt]))

    phrases: list[str] = []
    for segment in re.split(r"[,.!?]", lower_prompt):
        segment = segment.strip()
        if not segment:
            continue
        phrases.append(segment)
        words = segment.split()
        for index, word in enumerate(words):
            two = " ".join(words[index : index + 2]).strip()
            three = " ".join(words[index : index + 3]).strip()
            if two:
                phrases.append(two)
            if three:
                phrases.append(three)

    for candidate in phrases:
        filtered = filter_known_tags([candidate])
        if filtered:
            matches.add(filtered[0])

    avoid_parts = re.split(r"but not|avoid|no |don't want|not ", lower_prompt, flags=re.I)
    avoid_tags = (
        filter_known_tags(avoid_parts[1].split()) if len(avoid_parts) > 1 and avoid_parts[1] else []
    )

    constraints = IntentConstraints(
        **extract_numbers(prompt),
        single_player=True if re.search(r"single[\s-]?player", prompt, re.I) else None,
        multiplayer=True if re.search(r"multiplayer|co-op|coop|friends", prompt, re.I) else None,
    )

    return ParsedUserIntent(
        preferred_tags=[tag for tag in matches if tag not in avoid_tags],
        avoid_tags=avoid_tags,
        free_text_intent=prompt.strip(),
        constraints=constraints,
    )


def parse_user_intent(prompt: str) -> ParsedUserIntent:
    settings = get_settings()
    if not settings.openai_api_key:
        return heuristic_intent(prompt)

    model = ChatOpenAI(
        api_key=settings.openai_api_key,
        model="gpt-5.4-mini",
        temperature=0,
    )
    structured_model = model.with_structured_output(ParsedUserIntent)
    result = structured_model.invoke(
        [
            (
                "system",
                "You extract Steam search intent. Only include tags that are exact Steam tags. Put non-tag nuance in free_text_intent.",
            ),
            ("human", prompt),
        ]
    )

    return ParsedUserIntent(
        preferred_tags=filter_known_tags(result.preferred_tags),
        avoid_tags=filter_known_tags(result.avoid_tags),
        free_text_intent=result.free_text_intent or prompt.strip(),
        constraints=result.constraints,
    )
